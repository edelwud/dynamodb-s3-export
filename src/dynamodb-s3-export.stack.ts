import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, Table } from "aws-cdk-lib/aws-dynamodb";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import { Rule, Schedule } from "aws-cdk-lib/aws-events";
import { LambdaFunction } from "aws-cdk-lib/aws-events-targets";
import { LayerVersion, Tracing } from "aws-cdk-lib/aws-lambda";
import { NodejsFunctionProps } from "aws-cdk-lib/aws-lambda-nodejs";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { ExportFunction } from "./services/export-function";

export interface DynamoDBS3ExportStackProps extends StackProps {
  vpc?: Vpc;
}

export class DynamoDBS3ExportStack extends Stack {
  lambdaCommonProps: Partial<NodejsFunctionProps> = {
    vpc: this.props?.vpc,
    tracing: Tracing.ACTIVE,
    timeout: Duration.seconds(30),
    environment: {
      NODE_OPTIONS: "--enable-source-maps",
      LOG_LEVEL: "DEBUG",
      POWERTOOLS_METRICS_NAMESPACE: "DDBToS3Export",
    },
    layers: [
      LayerVersion.fromLayerVersionArn(
        this,
        "powertools-layer",
        `arn:aws:lambda:${
          Stack.of(this).region
        }:094274105915:layer:AWSLambdaPowertoolsTypeScript:9`
      ),
    ],
  };

  dataTable = new Table(this, "DataTable", {
    partitionKey: {
      name: "name",
      type: AttributeType.STRING,
    },
    pointInTimeRecovery: true,
  });

  destinationBucket = new Bucket(this, "Destination", {
    enforceSSL: true,
  });

  exportLambda = new ExportFunction(
    this,
    "ExportFunction",
    this.lambdaCommonProps
  );

  constructor(
    scope: Construct,
    id: string,
    private props?: DynamoDBS3ExportStackProps
  ) {
    super(scope, id, props);

    new Rule(this, "ExportInitiator", {
      schedule: Schedule.cron({
        minute: "0",
        hour: "2",
      }),
      targets: [new LambdaFunction(this.exportLambda)],
    });

    this.dataTable.grantReadData(this.exportLambda);
    this.destinationBucket.grantPut(this.exportLambda, "*.csv");

    this.exportLambda.addEnvironment(
      "DATA_TABLE_NAME",
      this.dataTable.tableName
    );
    this.exportLambda.addEnvironment(
      "DESTINATION_BUCKET_NAME",
      this.destinationBucket.bucketName
    );
    this.exportLambda.addEnvironment(
      "POWERTOOLS_SERVICE_NAME",
      "ExportProcess"
    );
  }
}
