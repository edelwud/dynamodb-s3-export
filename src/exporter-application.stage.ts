import { Stage, StageProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { DynamoDBS3ExportStack } from "./dynamodb-s3-export.stack";
import { NetworkStack } from "./network.stack";

export class ExporterApplicationStage extends Stage {
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    const { vpc } = new NetworkStack(this, "network");
    new DynamoDBS3ExportStack(this, "dynamodb-s3-export", { vpc });
  }
}
