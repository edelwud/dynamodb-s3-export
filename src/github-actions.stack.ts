import { Stack, StackProps } from "aws-cdk-lib";
import { ShellStep } from "aws-cdk-lib/pipelines";
import { NagSuppressions } from "cdk-nag";
import {
  AwsCredentials,
  GitHubActionRole,
  GitHubWorkflow,
} from "cdk-pipelines-github";
import { Construct } from "constructs";

export class GithubActionsStack extends Stack {
  role = new GitHubActionRole(this, "GithubActionsRole", {
    repos: ["edelwud/dynamodb-s3-export"],
  });

  pipeline = new GitHubWorkflow(this, "Pipeline", {
    synth: new ShellStep("Build", {
      commands: ["yarn", "npx projen build"],
    }),
    awsCreds: AwsCredentials.fromOpenIdConnect({
      gitHubActionRoleArn: `arn:aws:iam::${this.account}:role/GitHubActionRole`,
      roleSessionName: "github-actions-runner",
    }),
  });

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    NagSuppressions.addStackSuppressions(this, [
      {
        id: "AwsSolutions-IAM5",
        reason: "Allow to create all resources for deploy role",
        appliesTo: ["Resource::*"],
      },
    ]);
  }
}
