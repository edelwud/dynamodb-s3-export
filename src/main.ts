import { PDKNag } from "@aws-prototyping-sdk/pdk-nag";
import { ExporterApplicationStage } from "./exporter-application.stage";
import { GithubActionsStack } from "./github-actions.stack";

const app = PDKNag.app();

const env = {
  account: "762714406455",
  region: "us-east-1",
};

const githubStack = new GithubActionsStack(app, "GithubActions", { env });

const exporterAppProd = new ExporterApplicationStage(app, "prod", { env });

const exporterAppDev = new ExporterApplicationStage(app, "dev", { env });

githubStack.pipeline.addStageWithGitHubOptions(exporterAppProd, {});
githubStack.pipeline.addStage(exporterAppDev);

app.synth();
