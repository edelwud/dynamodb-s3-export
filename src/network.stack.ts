import { Stack } from "aws-cdk-lib";
import { GatewayVpcEndpointAwsService, Vpc } from "aws-cdk-lib/aws-ec2";

export class NetworkStack extends Stack {
  vpc = new Vpc(this, "VPC", {
    natGateways: 0,
    maxAzs: 3,
    gatewayEndpoints: {
      S3: {
        service: GatewayVpcEndpointAwsService.S3,
      },
      DDB: {
        service: GatewayVpcEndpointAwsService.DYNAMODB,
      },
    },
  });
}
