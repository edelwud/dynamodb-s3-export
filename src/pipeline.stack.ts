import { PDKPipeline } from "@aws-prototyping-sdk/pipeline";
import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";

export class PipelineStack extends Stack {
  readonly pipeline: PDKPipeline;

  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    this.pipeline = new PDKPipeline(this, "ApplicationPipeline", {
      primarySynthDirectory: "cdk.out",
      repositoryName:
        this.node.tryGetContext("repositoryName") || "ddb-s3-export",
      defaultBranchName: "main",
      publishAssetsInParallel: false,
      crossAccountKeys: true,
      synth: {},
      sonarCodeScannerConfig: this.node.tryGetContext("sonarqubeScannerConfig"),
    });
  }
}
