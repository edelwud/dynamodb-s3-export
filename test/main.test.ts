import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { DynamoDBS3ExportStack } from "../src/main";

test("Snapshot", () => {
  const app = new App();
  const stack = new DynamoDBS3ExportStack(app, "test");

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
