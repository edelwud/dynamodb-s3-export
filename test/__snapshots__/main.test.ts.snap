// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DataTable447BC44E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "name",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "name",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "Destination920A3C57": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "DestinationPolicy7982387E": {
      "Properties": {
        "Bucket": {
          "Ref": "Destination920A3C57",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "Destination920A3C57",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "Destination920A3C57",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "ExportFailureQueue52DA2A96": {
      "DeletionPolicy": "Delete",
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ExportFunctionAlarm9296DABF": {
      "Properties": {
        "AlarmDescription": "There are messages in the Dead Letter Queue",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "QueueName",
            "Value": {
              "Fn::GetAtt": [
                "ExportFailureQueue52DA2A96",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ApproximateNumberOfMessagesVisible",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ExportFunctionD1E50CEC": {
      "DependsOn": [
        "ExportFunctionServiceRoleDefaultPolicy3C842E61",
        "ExportFunctionServiceRole0CD064F7",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "0c36b9cd29a312cd99fa36b507777608cd9ebcee856cf914477ae47ce7f87469.zip",
        },
        "Description": "src/services/export.lambda.ts",
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DATA_TABLE_NAME": {
              "Ref": "DataTable447BC44E",
            },
            "DESTINATION_BUCKET_NAME": {
              "Ref": "Destination920A3C57",
            },
            "LOG_LEVEL": "DEBUG",
            "NODE_OPTIONS": "--enable-source-maps",
            "POWERTOOLS_METRICS_NAMESPACE": "DDBToS3Export",
            "POWERTOOLS_SERVICE_NAME": "ExportProcess",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:lambda:",
                {
                  "Ref": "AWS::Region",
                },
                ":094274105915:layer:AWSLambdaPowertoolsTypeScript:9",
              ],
            ],
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "ExportFunctionServiceRole0CD064F7",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "ExportFunctionEventInvokeConfig0CBBC94B": {
      "Properties": {
        "DestinationConfig": {
          "OnFailure": {
            "Destination": {
              "Fn::GetAtt": [
                "ExportFailureQueue52DA2A96",
                "Arn",
              ],
            },
          },
        },
        "FunctionName": {
          "Ref": "ExportFunctionD1E50CEC",
        },
        "Qualifier": "$LATEST",
      },
      "Type": "AWS::Lambda::EventInvokeConfig",
    },
    "ExportFunctionServiceRole0CD064F7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExportFunctionServiceRoleDefaultPolicy3C842E61": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ExportFailureQueue52DA2A96",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "DataTable447BC44E",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": [
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "Destination920A3C57",
                        "Arn",
                      ],
                    },
                    "/*.csv",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ExportFunctionServiceRoleDefaultPolicy3C842E61",
        "Roles": [
          {
            "Ref": "ExportFunctionServiceRole0CD064F7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ExportInitiator4CD48866": {
      "Properties": {
        "ScheduleExpression": "cron(0 2 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ExportFunctionD1E50CEC",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ExportInitiatorAllowEventRuletestExportFunction4551E33CC65C171F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ExportFunctionD1E50CEC",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ExportInitiator4CD48866",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
